fix: GCS endpoint validation to allow ports

ammarhusain-sol opened the pull request:
fix: Fixed the endpoint validation issue for GCS if port number is added to the URL

harshavardhana reviewed (changes requested):
why do you add a port number? for perfectly simple URL like 

```
storage.googleapis.com
```
?

ammarhusain-sol commented:
In the case of a URL with added port number, the validation check fails even though the endpoint is `storage.googleapis.com`

harshavardhana commented:
> In the case of a URL with added port number, the validation check fails even though the endpoint is `storage.googleapis.com`

yes but why add the port, `http` is port 80 and `https` is port 443, why specify it again?

ammarhusain-sol commented:
You are saying by default http and https requests are handled through 80 and 443 so why do we have to mention it explicitly that is what you are asking right?
The reason I have added this here is that restricting the users not to add port number in the request endpoint I thought of as a limitation so I added also I faced this problem only for the GCS endpoint alone not for the AWS S3 endpoint, that is another reason why I added it.

harshavardhana commented on pkg/s3utils/utils.go:
```diff
@@ -31,6 +31,7 @@ import (
 
 // Sentinel URL is the default url value which is invalid.
 var sentinelURL = url.URL{}
+var host string
```
this is wrong - please remove

harshavardhana commented on pkg/s3utils/utils.go:
```diff
@@ -223,10 +224,11 @@ func IsAmazonPrivateLinkEndpoint(endpointURL url.URL) bool {
 		return false
 	}
-	return endpointURL.Host == "storage.googleapis.com"
+	return host == "storage.googleapis.com"
```
```suggestion
	return endpointURL.Hostname() == "storage.googleapis.com"
```

harshavardhana commented on pkg/s3utils/utils.go:
```diff
@@ -409,3 +411,13 @@ func CheckValidObjectName(objectName string) error {
 	return CheckValidObjectNamePrefix(objectName)
 }
+
+//checks for port number
+func HasPortNumber(endpointURL url.URL) string {
+	if colonIndex := strings.LastIndex(endpointURL.Host, ":"); colonIndex != -1 {
+		host = endpointURL.Host[:colonIndex]
+	} else {
+		host = endpointURL.Host
+	}
+	return host
+}
```
this is not needed

harshavardhana commented:
> You are saying by default http and https requests are handled through 80 and 443 so why do we have to mention it explicitly that is what you are asking right?
> The reason I have added this here is that restricting the users not to add port number in the request endpoint I thought of as a limitation so I added also I faced this problem only for the GCS endpoint alone not for the AWS S3 endpoint, that is another reason why I added it.

I added review comments, we can take it in with the requested changes.

ammarhusain-sol commented:
@harshavardhana I made the changes.

harshavardhana commented:
> @harshavardhana I made the changes.

Please run gofumpt and goimports

harshavardhana commented on go.mod:
```diff
@@ -9,15 +9,18 @@ require (
-	golang.org/x/net v0.23.0
+	golang.org/x/crypto v0.25.0
+	golang.org/x/net v0.27.0
 	gopkg.in/ini.v1 v1.67.0
 )
 
 require (
 	github.com/davecgh/go-spew v1.1.1 // indirect
 	github.com/klauspost/cpuid/v2 v2.2.6 // indirect
 	github.com/stretchr/testify v1.3.0 // indirect
-	golang.org/x/sys v0.18.0 // indirect
-	golang.org/x/text v0.14.0 // indirect
+	golang.org/x/mod v0.19.0 // indirect
+	golang.org/x/sync v0.7.0 // indirect
+	golang.org/x/sys v0.22.0 // indirect
+	golang.org/x/text v0.16.0 // indirect
+	golang.org/x/tools v0.23.0 // indirect
 )
```
do not update `go mod` if you don't intend to, please revert these changes.

ammarhusain-sol replied:
I reverted those changes sorry for the overhead.

vadmeste reviewed (approved):
LGTM

klauspost commented:
Test is failing:
```
2024-07-11T18:04:06.2797859Z ##[error]    utils_test.go:136: Test 18: Expected to fail with <ERROR> "Google Cloud Storage endpoint should be 'storage.googleapis.com'.", but passed instead
2024-07-11T18:04:06.2808886Z --- FAIL: TestGetEndpointURL (0.00s)
```

ammarhusain-sol commented:
@klauspost found the issue in test18 it has been written that the GCP endpoint with a port number should fail, since we made that change it is not failing.
Thanks for sharing

harshavardhana commented:
FWIW I fixed it locally, waiting for the build will be merged.
